python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,486 @@
+
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    async def test_pubsub_channel_boundary_pattern(self, request, cluster_mode):
+        """ Tests that messages do not cross channel boundaries by pattern subsctiption """
+
+        CLIENTS_COUNT_PER_CHANNEL = 5
+        CHANNEL_NAMES = {"test-channel-1", "test-channel-2"}
+        CHANNEL_PATTERNS = {"test?channel-1", "test?channel-2"}
+        MESSAGES_BY_CHANNEL = {
+            channel : {
+                id : {
+                    channel: {
+                        "on_" + channel + "_" + msg + str(id): 1 for msg in ["foo_from_", "bar_from_", "baz_from_"]
+                    }
+                } for id in range(CLIENTS_COUNT_PER_CHANNEL)
+            } for channel in CHANNEL_NAMES
+        }
+
+        all_clients: List[TestPubSub.PubSubClient] = []
+        for channel_id in len(CHANNEL_NAMES):
+            for id in range(CLIENTS_COUNT_PER_CHANNEL):
+                all_clients.append(TestPubSub.PubSubClient(
+                    subsciptions={
+                        CoreCommands.ChannelModes.Pattern: {CHANNEL_PATTERNS[channel_id]}
+                    },
+                    sharded_messages_to_publish=[],
+                    unsharded_messages_to_publish=MESSAGES_BY_CHANNEL[CHANNEL_NAMES[channel_id]][id],
+                    cluster_mode=cluster_mode,
+                    request=request))
+                all_clients[-1].connect_and_subscribe()
+        
+        TestPubSub.publish_and_assert(all_clients)
+
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("sharded_pubsub", [True, False])
+    async def test_pubsub_callbacks_per_channel(self, request, cluster_mode, sharded_pubsub):
+        """ Tests that callbacks are per channel """
+
+        CHANNEL_NAME = "test-channel-1"
+        CLIENTS_COUNT = 1
+
+        callback_a_called = False
+        callback_b_called = False
+
+        def callback_a(message: CoreCommands.PubSubMsg, channel: str) -> None:
+            callback_a_called = True
+
+        def callback_b(message: CoreCommands.PubSubMsg, channel: str) -> None:
+            callback_b_called = True
+
+        redis_client = await create_client(self.request, self.cluster_mode)
+        await redis_client.subscribe(channels_or_patterns={"test-channel-1"},
+                                     channel_mode=CoreCommands.ChannelModes.Exact,
+                                     callback=callback_a,
+                                     context=None)
+        
+        await redis_client.subscribe(channels_or_patterns={"test-channel-2"},
+                                     channel_mode=CoreCommands.ChannelModes.Exact,
+                                     callback=callback_b,
+                                     context=None)
+        
+        await redis_client.publish(message="hi", channels={"test-channel-1"}, sharded=sharded_pubsub)
+        # FIXME - wait for receive with TO
```

Comment:
take a look on the init_future and init_callback in create function in redis_client.py file, you can use this pattern

Reply:
i will use a simple async sleep to allow the backend to complete the sending

Reply:
done